---
title: 'Assignment #4 Dworkin'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r}
#Bioconductor  version

#BiocVersion() from BiocInstaller
#First, check the version using the syntax package::function().
#Then, check the version using only the name of the function().

# Load the BiocInstaller package
library(BiocInstaller)

# Explicit syntax to check the Bioconductor version
BiocInstaller::biocVersion()

# When BiocInstaller is loaded use biocVersion alone
biocVersion()
```

```{r}
# Load the BSgenome package
library(BSgenome)

# Check the version of the BSgenome package
packageVersion("BSgenome")
```

```{r}
#S4class definition
showClass("className")
```


```{r}
#Interaction with classes

#Let's say we have an object called a_genome, from class BSgenome. With a_genome, you can ask questions like these:

# Check its main class
#class(a_genome)  # "BSgenome"

# Check its other classes
#is(a_genome)  # "BSgenome", "GenomeDescription"

# Is it an S4 representation?
#isS4(a_genome)  # TRUE
#If you want to find out more about the a_genome object, you can use the accessor show(a_genome) or use other specific accessors, from the list of .S4methods(class = "BSgenome").

# Investigate about the a_genome using show()
show(a_genome)

# Investigate some other accesors, has to use one per line
organism(a_genome)
provider(a_genome)
seqinfo(a_genome)
```

```{r}
#Discovering the Yeast genome

# Load the yeast genome
library(BSgenome.Scerevisiae.UCSC.sacCer3)

# Assign data to the yeastGenome object
yeastGenome <- BSgenome.Scerevisiae.UCSC.sacCer3

# Get the head of seqnames and tail of seqlengths for yeastGenome
head(seqnames(yeastGenome))
tail(seqlengths(yeastGenome))

# Select chromosome M, alias chrM
yeastGenome$chrM

# Count characters of the chrM sequence
nchar(yeastGenome$chrM)

```

```{r}
#Partitioning the Yeast genome
#getSeq example : getSeq(yeastGenome, names = "chrI", start = 100, end = 150)
#names is optional. if not specified it will return all chromosomes. parameters start and end are optional, you could provide one or both.

# Load the yeast genome
library(BSgenome.Scerevisiae.UCSC.sacCer3)

# Assign data to the yeastGenome object
yeastGenome <- BSgenome.Scerevisiae.UCSC.sacCer3

# Get the first 30 bases of each chromosome
getSeq(yeastGenome, end = 30)

```

```{r}
# available.genomes()
```

```{r}
#Exploring the Zika virus sequence

alphabet() #shows the letters included in the sequence
alphabetFrequency() # shows the counts per letter
#each alphabet corresponds to a specific biostring container, and each alphabet usually have extra code letters and symbols

# Load packages
library(Biostrings)

# Check the alphabet of the zikaVirus
alphabet(zikaVirus)

# Check the alphabetFrequency of the zikaVirus
alphabetFrequency(zikaVirus)

# Check alphabet of the zikaVirus using baseOnly = TRUE
alphabet(zikaVirus, baseOnly = TRUE)

# by putting baseOnly = TRUE, it only outputs the alphabet letter that are nucleotide bases


```

```{r}
#Manipulating Biostrings

# Unlist the set and select the first 21 letters as dna_seq, then print it
dna_seq <- subseq(unlist(zikaVirus), end = 21)
dna_seq

# 1.1 Transcribe dna_seq as rna_seq, then print it
rna_seq <- RNAString(dna_seq) 
rna_seq
#notice to transcribe, you gotta use RNAString

# 1.2 Translate rna_seq as aa_seq, then print it
aa_seq <- translate(rna_seq)
aa_seq

# 2.1 Translate dna_seq as aa_seq_2, then print it
aa_seq_2 <- translate(dna_seq)
aa_seq_2

```

```{r}
#From a set to a single sequence
# Create zikv with one collated sequence using `zikaVirus`
zikv <- subseq(unlist(zikaVirus))

# Check the length of zikaVirus and zikv
length(zikaVirus)
length(zikv)

# Check the width of zikaVirus
width(zikaVirus)

# Subset zikv to only the first 30 bases
subZikv <- subseq(zikv, end = 30)
subZikv
```

```{r}
#Subsetting a set
subseq(zikaSet, 
        start = c(20, 40, 2), 
        end = c(50, 45, 22)
     )
```

```{r}
#Common sequence manipulation functions
`reverse()` 
`complement()`
`reverseComplement()`
`translate()

# The reverse of zikv is
reverse(zikv)

# The complement of zikv is
complement(zikv)

# The reverse complement of zikv is
reverseComplement(zikv)

# The translation of zikv is
translate(zikv)
```

```{r}
#Searching for a pattern
#Let's find some occurrences of a pattern in the zikaVirus set using vmatchPattern(). Then, let's try the same pattern search using matchPattern() with a single sequence zikv.

# For Sets
#vmatchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
             # subject = zikaVirus, max.mismatch = 1)
# For single sequences
#matchPattern(pattern = "ACATGGGCCTACCATGGGAG", 
             # subject = zikv, max.mismatch = 1)
#Both functions should find the same number of occurrences, but you will notice a different output. How many matches do we get when running each pattern search individually?


```

```{r}
#Finding Palindromes
# Find palindromes in zikv
findPalindromes(zikv)
```

```{r}
#Finding a conserved region within six frames
#Now you will be able to look for the NS5 protein sequence in the Zika virus sequence.

#The NS5 is a very conserved virus protein. It was downloaded and loaded for you from Uniprot.
#The Zika virus DNA sequence has been transcribed into a RNAStringSet, called rnaframesZikaSet. The set has six reading frames (one per sequence), for you to translate into amino acids.
#When you do the search, allow max.mismatch = 15 to add flexibility to your search.


```

```{r}
#Finding a conserved region within six frames
# print the rnaframesZikaSet 
rnaframesZikaSet

# translate all 6 reading frames 
AAzika6F <- translate(rnaframesZikaSet)
AAzika6F

# Count the matches allowing 15 mistmatches
vcountPattern(pattern = ns5, subject = AAzika6F, max.mismatch = 15)

# Select the frame that contains the match
selectedSet <- AAzika6F[3]

#Convert this frame into a single sequence
selectedSeq <- unlist(selectedSet)
```

```{r}
#Looking for a match
#have to use matchPattern() when 'subject' is a XString object (single sequence)

# Use vmatchPattern with the set
vmatchPattern(pattern = ns5, subject = selectedSet, max.mismatch = 15)

# Use matchPattern with the single sequence
matchPattern(pattern = ns5, subject = selectedSeq, max.mismatch = 15)

```

```{r}
#IRanges
#its function IRanges(), which is a vector representation of a sequence range used to facilitate subsetting and annotation
# IRanges objects can be defined using either by numeric or logical vectors. Numeric vectors will delimit the exac positions by index, and logical vectors will select a range by condition.
```

```{r}
#Constructing IRanges
#Using the IRanges() function, you can specify some parameters, such as: start, end or width. These parameters can be numeric vectors, or you could set the start as a logical vector. Missing arguments will be resolved following the equation width = end - start + 1.

#The IRanges() constructor indicates that all of the parameters are optional with default NULL:

#IRanges(start = NULL, end = NULL, width = NULL, names = NULL)

# load package IRanges
library(IRanges)

# start vector 1 to 5, end 100 
IRnum1 <- IRanges(start = 1:5, end = 100)

# end 100, width 89 and 10
IRnum2 <- IRanges(end = 100, width = c(89, 10))

# logical argument start = Rle(c(F, T, T, T, F, T, T, T))
IRlog1 <- IRanges(start = Rle(c(F, T, T, T, F, T, T, T)))

# Printing objects in a list
print(list(IRnum1 = IRnum1, IRnum2 = IRnum2, IRlog1 = IRlog1))

```

```{r}
#Interacting with IRanges
#You can use the IRanges() function to create a single sequence:

# First example
#(IRanges(start = 10, end = 37))
But also you can use vectors to create multiple sequence ranges at the same time. This is both fascinating and useful!

# second example
#(IRanges(start = c(5, 35, 50),
                  #end = c(12, 39, 61),
            # names = LETTERS[1:3]))
#Remember width = end - start + 1.

#For this exercise, check the lengths of each of the examples provided here. Use either width() or lengths() functions with each of the examples and see what the difference is before selecting an answer .
```

```{r}
#GenomicRanges accessors

#In the previous exercise, you created a GRanges object from a data.frame with the basic information. You will discover that GRanges can store much more!

#Use the GRanges object called myGR to investigate more of it using accessors methods. You can check the GRanges object's characteristics like the name of each chromosome, the number of sequences, the names of each sequence, information about strand, score, length and more.

#The following are some of the basic accessors for GRanges:

#seqnames(gr)
#ranges(gr)
#mcols(gr)
#genome(gr)
#seqinfo(gr)
#For a complete list of accessors, you can check methods(class = "GRanges").

#Use the object myGR for this exercise.

# Load Package Genomic Ranges
library(GenomicRanges)

# Print the GRanges object myGR
myGR

# Check the metadata, if any
mcols(myGR)

```

```{r}
#ABCD1 mutation
#You have just learned about the gene ABCD1. It encodes the protein in charge of the normal transport of fats that keep brain and lung cells, functioning normally. When these groups of fats are not broken down, they build up in the body and become toxic. This affects the adrenal glands (small glands on top of each kidney) and the insulation (myelin) that surrounds neurons, causing hormonal problems, deteriorating movement, vision and hearing. More than 650 mutations in the ABCD1 gene have been found to cause X-linked adrenoleukodystrophy, a rare genetic disease.

#Since you are going to be studying this gene in the coming exercises, it is important to remember where is located. If you are unsure, check the gene ABCD1 and its location using Ensembl genome browser.


```

```{r}
#Human genome chromosome X

#It is your turn to use the TxDb.Hsapiens.UCSC.hg38.knownGene package and extract information from it. Like in the video, you will subset all genes in chromosome X, using the function genes() with filter tx_chrom = "chrX". Then, you will explore this subset of genes.

#Remember that filter receives a list() of filter conditions, to select specific genome intervals.

#As extra information if you would like to test other filters, valid names for this list are: "gene_id", "tx_id", "tx_name", "tx_chrom", "tx_strand", "exon_id", "exon_name", "exon_chrom", "exon_strand", "cds_id", "cds_name", "cds_chrom", "cds_strand" and "exon_rank"

# load human reference genome hg38
library(TxDb.Hsapiens.UCSC.hg38.knownGene)

# assign hg38 to hg, then print it
hg <- TxDb.Hsapiens.UCSC.hg38.knownGene
hg

# filter1: extract all the genes in chromosome X as hg_chrXg, then print it
hg_chrXg <- genes(hg, filter = list(tx_chrom = c("chrX")))
hg_chrXg

# filter2: extract all positive stranded genes in chromosome X as hg_chrXgp, then sort it
hg_chrXgp <- genes(hg, filter = list(tx_chrom = c("chrX"), tx_strand = "+"))
sort(hg_chrXgp)

```

```{r}
#A sequence window
To split your GRanges

#slidingWindows() function where you specify arguments such as width and step.

#slidingWindows(x, width, step = 1L)

#If you wanted exactly 2 windows using step = 1L, what will be the maximum number allowed for the width of a window?'

#The length of the GRanges object -1
```

```{r}
#Is it there?
#You can find accessors using the function methods(class = "GRangesList")
#It is your turn to explore chromosome X genes hg_chrX, and find the gene of interest ABCD1. Using the function overlapsAny() between the target ABCD1 and the subject hg_chrX

overlapsAny(ABCD1, hg_chrX)

can use table to count 
```

```{r}
#More about ABCD1
# Store the overlapping range in rangefound
rangefound <- subsetByOverlaps(hg_chrX, ABCD1)

# Check names of rangefound
names(rangefound)

# Check the geneOfInterest 
ABCD1

# Check rangefound
rangefound
```

```{r}
#How many transcripts?

# load the human transcripts DB to hg
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
hg <- TxDb.Hsapiens.UCSC.hg38.knownGene

# prefilter chromosome X
seqlevels(hg) <- c("chrX")

# get all transcripts by gene and print it
hg_chrXt <- transcriptsBy(hg, by = "gene")
hg_chrXt

# select gene `215` from the transcripts
hg_chrXt$`215`

```

```{r}
#From GRangesList object into a GRanges object


```

```{r}
#Why fastq?
#fastq formated fileds are not only the standard for high-throughput sequence, They are also used to save the score encodings for each based called which helps to assess the quality of each and all of the reads in a sequence file.
```

```{r}
#Exploring a fastq file

# load ShortRead
library(ShortRead)

# load ShortRead
library(ShortRead)

# print fqsample
fqsample

# class of fqsample
class(fqsample)

# class sread fqsample
class(sread(fqsample))

# id fqsample
id(fqsample)
```

```{r}
# FastqSampler() draws a subsample from a fastq file and yield() is the function used to extract the sample from the FastqSmapler
```

```{r}
#Exploring sequence quality

# load ShortRead
library(ShortRead)

# Check quality
quality(fqsample)

# Check encoding
encoding(quality(fqsample))

# Check baseQuality
qaSummary[["baseQuality"]]

```

```{r}
#Try your own nucleotide frequency plot
#Look at the frequency of nucleotides per cycle - make a visualization. 

#Now it's time to take a closer look at the frequency of nucleotides per cycle. The best way to do this is by making a visualization. Usually the first cycles are a bit random and then the frequency of nucleotides should stabilize with the coming cycles.

#This exercise uses the complete fastq file SRR1971253 with some pre processing done for you.

#library(ShortRead)
#fqsample <- readFastq(dirPath = "data", 
#                      pattern = "SRR1971253.fastq")
#abc <- alphabetByCycle(sread(fqsample))
# transpose nucleotides A, C, G, T per column
#nucByCycle <- t(abc[1:4,]) 
# tidy dataset
#nucByCycle <- nucByCycle %>% 
#  as.tibble() %>% # convert to tibble
#  mutate(cycle = 1:50) # add cycle numbers

#Your task is to make a Nucleotide Frequency by Cycle plot, using tidyverse functions!

#glimpse() the object nucByCycle, you will make an awesome plot with it!
#gather() the nucleotide letters in alphabet and get a new count column.
#Make a line plot of cycle vs count colored by alphabet.
#The rest is done for you, Cheers!

# glimpse nucByCycle
glimpse(nucByCycle)

# make an awesome plot!
nucByCycle %>% 
  # gather the nucleotide letters in alphabet and get a new count column
  gather(key = alphabet, value = count , -cycle) %>% 
  ggplot(aes(x = cycle, y =  count, color = alphabet)) +
  geom_line(size = 0.5 ) +
  labs(y = "Frequency") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank())
```

```{r}
#Filtering reads on the go!

#eg. you are interested only in those reads that start with the pattern "ATGCA". can use the function srFilter()

#myStartFilter <- srFilter(function(x) substr(sread(x), 1, 5) == "ATGCA")

# Load package ShortRead
library(ShortRead)

# Check class of fqsample
class(fqsample)

# Filter reads into selectedReads using myStartFilter
selectedReads <- fqsample[myStartFilter(fqsample)]

# Check class of selectedReads
class(selectedReads)

# Check detail of selectedReads
detail(selectedReads)
```

```{r}
#Removing duplicates
# sample with duplicates of class: ShortReadQ
dfqsample

# get the reads from dfqsample
mydReads <- sread(dfqsample)

# Counting duplicates
table(srduplicated(mydReads))

# mydReads[srduplicated(mydReads) == FALSE]
# if TRUE, it will return those reads that are duplicated, but we want to KEEP those that are not duplicated.
```

```{r}
#More filtering!

```

```{r}
#PLotting cycle average quality
# Load package Rqc
library(Rqc)

# Average per cycle quality plot
rqcCycleAverageQualityPlot(qa)

# Average per cycle quality plot with white background
rqcCycleAverageQualityPlot(qa) + theme_minimal()

# Read quality plot with white background
rqcReadQualityPlot(qa) + theme_minimal()

#By adding theme_minimal() it makes the background white
```

















